### ISSUE TITLE
Add requestFactory as an argument to Site.__init__

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7016]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#7016|
|Type|enhancement|
|Created|2014-03-03 11:37:02Z|
|Branch|https://github.com/twisted/twisted/tree/site-requestfactory-7016-3|

From [#6933](https://github.com/twisted/twisted/issues/6933)

            4. Customising the session name can then be achieved by assigning a
               curried Request constructor to server.Site.requestFactory eg
               mysite.requestFactory = functools.partial(Request,
               sessionCookieBaseName='MYSITE_SESSION')
               1. It would be even better if server.Site accepted a
     "requestFactory"
                  argument, but that's another ticket.

Attachments:

* [7016-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/66c/66cda3fdb29490e5bd0e880aa1a76d7679b1f90d/c252eec8f6ddcb267d97147bd581d48d21cd2606.diff) (3118 bytes) - added by adiroiban on 2014-04-22 10:19:27Z - 
* [7016-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/66c/66cda3fdb29490e5bd0e880aa1a76d7679b1f90d/756f5cf2c73d2e1b0ca6f3b5dd2e93f78186ab3e.diff) (3096 bytes) - added by adiroiban on 2014-11-10 20:26:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7016 7016
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_site_requestfactory_7016_3 branches/site-requestfactory-7016-3
branch_author__adiroiban__rwall adiroiban, rwall
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1393846622000000 1393846622000000
changetime__1447972926672962 1447972926672962
version__None None
owner__ 
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

I have attached a patch.

I have also added a test for buildProtocol since it was missing and I wanted to check how requestFactory leaves the Site instance... as its final destination is the HTTPChannel.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

(In [43416]) Branching to site-requestfactory-7016.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj set owner to @adiroiban<br>@wallrj set status to `new`|
|-|-|

Thanks adiroban,

This looks useful.

Notes:
 * Applied 7016-1.diff from adiroban
 * Forced build: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/site-requestfactory-7016
 * I guess this shouldn't be merged until [#6933](https://github.com/twisted/twisted/issues/6933) is resolved...although it doesn't strictly depend on it.

Points:
 1. branches/site-requestfactory-7016/twisted/web/server.py
    1. The standard way to handle optional initialiser arguments is to use ``if requestFactory is None:`` and set the default inside the initialiser.
       For example, see how the optional ``reactor`` argument of ``twisted.web.server.Session`` is handled.

 2. branches/site-requestfactory-7016/twisted/web/test/test_web.py
    1. twistedchecker warning ``W9016: 79,0: Too many blank lines after docstring, found 2``
    2. ``test_constructorRequestFactory``
       1. ``customFactory`` could just be a dummy. ie ``dummyRequestFactory = object()`` and ``server.Site(..., requestFactory=dummyRequestFactory``
    3. ``test_buildProtocol``
       1. I found the docstring difficult to understand.
          Perhaps it could be restated as "L{server.Site.buildProtocol} returns a C{Channel} whose C{site} and C{requestFactory} attributes are assigned from the C{site} instance.

    4. Add a test to demonstrate that the default ``requestFactory`` is ``Request``.

 3. Custom session cookie documentation from [#6933](https://github.com/twisted/twisted/issues/6933) can be updated to use this ``requestFactory`` argument instead.

Please answer or address the numbered points above and submit a new patch against [branches/site-requestfactory-7016](branches/site-requestfactory-7016) .

Thanks again and look forward to being able to use this feature in my Twisted web servers.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban removed owner|
|-|-|

* Why is this ticked blocked by [#6933](https://github.com/twisted/twisted/issues/6933) ?

------

1. branches/site-requestfactory-7016/twisted/web/server.py - Done. Hope the fix is good.

2. branches/site-requestfactory-7016/twisted/web/test/test_web.py

    1. twistedchecker warning W9016: 79,0: Done

    2. test_constructorRequestFactory: True. Done!

    3. test_buildProtocol. Thanks. I left out `L{server.Site.buildProtocol}` as this is the target of the test and should be understood from the context.

3. A assume that cookie documentation from [#6933](https://github.com/twisted/twisted/issues/6933)  should be updated **after** this is merged.

4. Done. I wrote a tests for default value.

----------

I will send the patch based on your branch. Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [44001]) Branching to site-requestfactory-7016-2.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @adiroiban|
|-|-|

Hi adiroiban,

Sorry to say, the patch doesn't apply cleanly any more at *all*; all the hunks were rejected.  So please update this for current trunk.  I guess the good news is that Twisted is a fairly active project ;-).

However, for this to be a real review I wanted to look at the patch as well, so you have something to go on.

I think you were attempting to address rwall's previous comment:

> The standard way to handle optional initialiser arguments is to use if requestFactory is None: and set the default inside the initialiser. For example, see how the optional reactor argument of twisted.web.server.Session is handled.

by setting `Site.requestFactory` to be `None` at the class level.  However, I'm pretty sure what rwall was saying was simply to do the check with `is` instead of the implicit `.__nonzero__()` of putting the `requestFactory` argument as the condition of an `if` block.  Setting the class variable to `None` is, among other problems, an incompatible change, because `Site.requestFactory` is a public name (even if it is documented and used as an `@ivar`, which is why we now generally prefer to avoid having the class variable at all).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

(In [44098]) Branching to site-requestfactory-7016-3.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review!

I am confused about how to implement this but I have created a new branch and updated the code with what I understood from current feedback.

Sending the branch to buildbot.

Please check latest changes.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_12"></a>|@wsanchez commented|
|-|-|

Reviewed; go for merge.  Don't forget NEWS file.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set status to `closed`|
|-|-|

(In [44505]) Merge site-requestfactory-7016-3: Add requestFactory as an argument to Site.__init__.

Author: adiroiban
Reviewers: rwall, glyph, wsanchez
Fixes: [#7016](https://github.com/twisted/twisted/issues/7016)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince commented|
|-|-|

Unfortunately, this introduced an incompatible change. If people were passing positional arguments, this changed the meaning of the first one, which used to be `logFile`.
